### ISSUE TITLE
bundle-sfx and source maps

### ISSUE THREAD

**Issue by netheril96** [jspm/jspm-cli#1328]
When I call `jspm bundle-sfx` I get a single file that can execute on its own. However, the source map generated still require the original directory structure to be in place to work. So I cannot deploy that single bundled js and source map and be able to both execute and debug. Is there any work around? In case it isn't clear, can I have a self executing bundle with a source map that works only on their own?

**Comment by jackfranklin** [comment]
Hey,

Are you passing any options to `bundle-sfx` ? I think the `--source-maps-contents` flag might solve this for you. 

Let me know!

Jack

**Comment by netheril96** [comment]
This is the traceback I got after passing the option `--source-maps-contents`

```
err  TypeError: Expected operator before resources/main.js
         at parseExpression (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/systemjs-builder/lib/arithmetic.js:24:13)
         at exports.traceExpression (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/systemjs-builder/lib/arithmetic.js:120:20)
         at /Users/rsy/homebrew/lib/node_modules/jspm/node_modules/systemjs-builder/lib/builder.js:539:12
         at lib$rsvp$$internal$$tryCatch (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:493:16)
         at lib$rsvp$$internal$$invokeCallback (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:505:17)
         at /Users/rsy/homebrew/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:1001:13
         at lib$rsvp$asap$$flush (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:1198:9)
         at doNTCallback0 (node.js:428:9)
         at process._tickCallback (node.js:357:13)
         at Function.Module.runMain (module.js:459:11)
zsh: exit 1     jspm bundle-sfx --source-maps-contents resources/main.js js/boxes_bundle.js
```

I tried placing the flag `--source-maps-contents` differently, but the error was always thrown out. Besides, is there any documentation about this flag?

**Comment by jackfranklin** [comment]
No I don't think there is any documentation at the moment. I'll see if I can write some up.

TL;DR is that `--source-maps-contents` will output the contents of all the files it minifies into the sourcemap. I've been able to bundle before and using this has avoided the need to push the actual source files to the server, so this is rather odd. That error is also bizarre. Presuming placing the flag at the end doesn't fix it?

**Comment by netheril96** [comment]
@jackfranklin 

> Presuming placing the flag at the end doesn't fix it?

No.

```
[`--> jspm bundle-sfx resources/main.js js/boxes_bundle.js --source-maps-contents
     Building the single-file sfx bundle for resources/main.js js/boxes_bundle.js...

err  TypeError: Expected operator before js/boxes_bundle.js
         at parseExpression (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/systemjs-builder/lib/arithmetic.js:24:13)
         at exports.traceExpression (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/systemjs-builder/lib/arithmetic.js:120:20)
         at /Users/rsy/homebrew/lib/node_modules/jspm/node_modules/systemjs-builder/lib/builder.js:539:12
         at lib$rsvp$$internal$$tryCatch (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:493:16)
         at lib$rsvp$$internal$$invokeCallback (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:505:17)
         at /Users/rsy/homebrew/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:1001:13
         at lib$rsvp$asap$$flush (/Users/rsy/homebrew/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:1198:9)
         at doNTCallback0 (node.js:428:9)
         at process._tickCallback (node.js:357:13)
         at Function.Module.runMain (module.js:459:11)
zsh: exit 1     jspm bundle-sfx resources/main.js js/boxes_bundle.js --source-maps-contents
```

**Comment by jackfranklin** [comment]
Very odd. Which version of jspm are you on?

**Comment by netheril96** [comment]
@jackfranklin 0.16.15

**Comment by guybedford** [comment]
@netheril96 the option is `--source-map-contents` not `--source-maps-contents`.

**Comment by netheril96** [comment]
@guybedford 

OK, I successfully executed that command. But it is not what I want. It seems that with `--source-map-contents` `jspm` will simply inline the source map in the bundle. But to actually take advantage of the map, the original source files still has to be in place. What I want is a similar flag to `browserify --debug`, where the bundle itself contains all the original source files, and can be debugged without the original sources.

**Comment by guybedford** [comment]
The original paths structure must remain in tact for the source maps paths to work out for debugging back to the original sources. How else might this work otherwise?

**Comment by netheril96** [comment]
@guybedford 

I don't what voodo `browserify` does, but it does work in the way I describe. Here is a screenshot of how it works:

First I create the bundle with `--debug` option, then I remove the original source files. Notice that the `js` folder has nothing but the bundled file.

<img width="526" alt="screen shot 2015-11-30 at 21 14 19" src="https://cloud.githubusercontent.com/assets/836839/11472446/8f9783b6-97a7-11e5-9718-42a8f842fe98.png">

Then I browse the localhost (with a simple webserver written in Node.JS) in Chrome, 

<img width="1214" alt="screen shot 2015-11-30 at 21 14 38" src="https://cloud.githubusercontent.com/assets/836839/11472464/b5676b06-97a7-11e5-8c4a-ccef1ba346ee.png">

Notice how those source files magically appear in the folder `js`, even if they don't appear in the disk anywhere.

**Comment by guybedford** [comment]
@netheril96 the above is due to source map contents, which is exactly what `--source-map-contents` does in jspm.

**Comment by netheril96** [comment]
@guybedford 

The difference is that after I delete the original sources, with JSPM the mapped virtual source files are shown as empty, but full with browserify.

With JSPM:

<img width="1224" alt="screen shot 2015-11-30 at 21 25 20" src="https://cloud.githubusercontent.com/assets/836839/11472736/969fcc20-97a9-11e5-9f9d-60abf335514c.png">

With browserify:

<img width="1214" alt="screen shot 2015-11-30 at 21 29 28" src="https://cloud.githubusercontent.com/assets/836839/11472746/a7116e7e-97a9-11e5-828f-fc5c3856fa78.png">

**Comment by jackfranklin** [comment]
@netheril96 are you able to recreate in a smaller project? I'm happy to take a look - `source-map-contents` has worked fine for me in the past.

Can you take a look at the generated sourcemap and look for the `sourcemapContents` array, checking that the array does look like it's got the right stuff in it?

**Comment by netheril96** [comment]
@jackfranklin I haven't be able to reproduce the issue on a small code base yet. In my case, `sourcemapContents` does not exist. `sourcesContent` exists but is an array of all `null`s.

**Comment by jackfranklin** [comment]
Ah sorry, I mistyped the name. `sourcesContent` being entirely null is definitely a bug, although not sure how that's occurring at the moment...

**Comment by jackfranklin** [comment]
That might be a bug in systemjs/builder, but I'm not really sure at the moment...I'll see if I can recreate locally.

**Comment by aniranc** [comment]
Indeed looks like a systemjs/builder problem. I opened an issue here with more info: https://github.com/systemjs/builder/issues/427

**Comment by guybedford** [comment]
Closing to track at the builder issue above.
